File: skills/qrcode-skills/scripts/batch_decode.js
#!/usr/bin/env node
/**
 * 批量解码二维码 (Node.js 版)
 *
 * 用法:
 *   node scripts/batch_decode.js --input <文件> [--column <列>] [--output-txt <路径>]
 *
 * 输出 JSON 格式与 Python 版一致。
 */

const https = require("https");
const http = require("http");
const fs = require("fs");
const os = require("os");
const path = require("path");
const { URL } = require("url");

const API_ENDPOINT = "https://api.2dcode.biz/v1/read-qr-code";
const FAIL = "未解析到二维码";

// ── 参数解析 ──────────────────────────────────────────────

function parseArgs(argv) {
  const args = { input: null, column: null, outputTxt: null };
  for (let i = 2; i < argv.length; i++) {
    switch (argv[i]) {
      case "--input": args.input = argv[++i]; break;
      case "--column": args.column = argv[++i]; break;
      case "--output-txt": args.outputTxt = argv[++i]; break;
    }
  }
  return args;
}

// ── HTTP 工具 ──────────────────────────────────────────────

function httpGet(url) {
  return new Promise((resolve, reject) => {
    const mod = url.startsWith("https") ? https : http;
    mod.get(url, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location)
        return httpGet(res.headers.location).then(resolve, reject);
      const chunks = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => resolve(Buffer.concat(chunks)));
    }).on("error", reject);
  });
}

function downloadToTemp(url) {
  return new Promise((resolve, reject) => {
    const ext = path.extname(url.split("?")[0]) || ".png";
    const tmp = path.join(os.tmpdir(), `qr_${Date.now()}_${Math.random().toString(36).slice(2)}${ext}`);
    const mod = url.startsWith("https") ? https : http;
    const file = fs.createWriteStream(tmp);
    mod.get(url, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        file.close(); fs.unlinkSync(tmp);
        return downloadToTemp(res.headers.location).then(resolve, reject);
      }
      res.pipe(file);
      file.on("finish", () => { file.close(); resolve(tmp); });
    }).on("error", (e) => { file.close(); if (fs.existsSync(tmp)) fs.unlinkSync(tmp); reject(e); });
  });
}

// ── 解码函数 ──────────────────────────────────────────────

function isUrl(s) { return s.startsWith("http://") || s.startsWith("https://"); }

async function tryWechatQr(source) {
  try {
    const { scan } = await import("qr-scanner-wechat");
    const sharp = require("sharp");

    let imgPath = source;
    let tmpPath = null;
    if (isUrl(source)) {
      tmpPath = await downloadToTemp(source);
      imgPath = tmpPath;
    }
    try {
      const { data, info } = await sharp(imgPath)
        .ensureAlpha()
        .raw()
        .toBuffer({ resolveWithObject: true });
      const result = await scan({
        data: Uint8ClampedArray.from(data),
        width: info.width,
        height: info.height,
      });
      return result?.text || null;
    } finally {
      if (tmpPath && fs.existsSync(tmpPath)) fs.unlinkSync(tmpPath);
    }
  } catch {
    return null;
  }
}

async function decodeApiUrl(imageUrl) {
  try {
    const apiUrl = `${API_ENDPOINT}?file_url=${encodeURIComponent(imageUrl)}`;
    const buf = await httpGet(apiUrl);
    const data = JSON.parse(buf.toString());
    if (data.code === 0 && data.data?.contents?.length) return data.data.contents.join("; ");
    return null;
  } catch { return null; }
}

async function decodeApiFile(filePath) {
  return new Promise((resolve) => {
    try {
      const boundary = `----formdata${Date.now()}`;
      const filename = path.basename(filePath);
      const fileData = fs.readFileSync(filePath);
      const header = Buffer.from(`--${boundary}\r\nContent-Disposition: form-data; name="file"; filename="${filename}"\r\nContent-Type: application/octet-stream\r\n\r\n`);
      const footer = Buffer.from(`\r\n--${boundary}--\r\n`);
      const body = Buffer.concat([header, fileData, footer]);
      const url = new URL(API_ENDPOINT);
      const options = { hostname: url.hostname, path: url.pathname, method: "POST", headers: { "Content-Type": `multipart/form-data; boundary=${boundary}`, "Content-Length": body.length } };
      const req = https.request(options, (res) => {
        const chunks = [];
        res.on("data", (c) => chunks.push(c));
        res.on("end", () => { try { const d = JSON.parse(Buffer.concat(chunks).toString()); resolve(d.code === 0 && d.data?.contents?.length ? d.data.contents.join("; ") : null); } catch { resolve(null); } });
      });
      req.on("error", () => resolve(null));
      req.write(body); req.end();
    } catch { resolve(null); }
  });
}

async function decodeSingle(source) {
  const local = await tryWechatQr(source);
  if (local) return local;

  if (isUrl(source)) {
    const r = await decodeApiUrl(source);
    return r || FAIL;
  } else if (fs.existsSync(source)) {
    const r = await decodeApiFile(source);
    return r || FAIL;
  } else {
    const r = await decodeApiUrl(source);
    return r || FAIL;
  }
}

// ── 文件读写 ──────────────────────────────────────────────

const DECODE_KEYWORDS = ["url", "link", "image", "img", "图片", "链接", "网址", "二维码"];

function readTxt(fp) { return fs.readFileSync(fp, "utf-8").split("\n").map((l) => l.trim()).filter(Boolean); }

function parseCsvLine(line) {
  const result = []; let current = "", inQ = false;
  for (const ch of line) { if (ch === '"') inQ = !inQ; else if (ch === "," && !inQ) { result.push(current.trim()); current = ""; } else current += ch; }
  result.push(current.trim()); return result;
}

function readCsv(fp) { return fs.readFileSync(fp, "utf-8").replace(/^\uFEFF/, "").split("\n").filter(Boolean).map(parseCsvLine); }

function readExcel(fp) {
  const XLSX = require("xlsx");
  const wb = XLSX.readFile(fp);
  return XLSX.utils.sheet_to_json(wb.Sheets[wb.SheetNames[0]], { header: 1, defval: "" }).map((r) => r.map((c) => String(c ?? "")));
}

function resolveCol(headers, column) {
  if (column != null) {
    const num = parseInt(column);
    if (!isNaN(num) && num >= 0 && num < headers.length) return num;
    const low = String(column).toLowerCase().trim();
    for (let i = 0; i < headers.length; i++) { if (headers[i].trim().toLowerCase() === low) return i; }
    return null;
  }
  for (let i = 0; i < headers.length; i++) {
    const h = headers[i].trim().toLowerCase();
    for (const kw of DECODE_KEYWORDS) { if (h.includes(kw)) return i; }
  }
  return headers.length === 1 ? 0 : null;
}

async function processTxt(inputPath, outputTxt) {
  const lines = readTxt(inputPath);
  const results = [];
  let success = 0;
  for (const line of lines) {
    const d = await decodeSingle(line);
    results.push(d);
    if (d !== FAIL) success++;
  }
  const stem = path.basename(inputPath, path.extname(inputPath));
  const outPath = outputTxt || path.join(path.dirname(inputPath), `${stem}_decoded.txt`);
  fs.writeFileSync(outPath, results.join("\n"), "utf-8");
  return { total: lines.length, success, failed: lines.length - success, output_file: path.resolve(outPath), output_txt: path.resolve(outPath) };
}

async function processCsv(inputPath, column, outputTxt) {
  const rows = readCsv(inputPath);
  if (!rows.length) return { error: "CSV 文件为空" };
  const headers = rows[0];
  const dataRows = rows.slice(1);
  const colIdx = resolveCol(headers, column);
  if (colIdx == null) return { need_column: true, columns: headers, preview: rows.slice(0, 6), message: column == null ? "无法自动判断 URL 列，请指定 --column 参数" : `找不到列 '${column}'` };

  const decoded = [];
  let success = 0;
  for (const row of dataRows) {
    const url = (row[colIdx] || "").trim();
    const d = url ? await decodeSingle(url) : FAIL;
    decoded.push(d);
    if (d !== FAIL) success++;
  }

  if (outputTxt) {
    fs.writeFileSync(outputTxt, decoded.join("\n"), "utf-8");
    return { total: dataRows.length, success, failed: dataRows.length - success, output_file: path.resolve(inputPath), output_txt: path.resolve(outputTxt) };
  }

  headers.push("解码结果");
  dataRows.forEach((r, i) => r.push(decoded[i]));
  const csvContent = [headers, ...dataRows].map((r) => r.map((c) => `"${String(c).replace(/"/g, '""')}"`).join(",")).join("\n");
  fs.writeFileSync(inputPath, "\uFEFF" + csvContent, "utf-8");
  return { total: dataRows.length, success, failed: dataRows.length - success, output_file: path.resolve(inputPath), output_txt: null };
}

async function processExcel(inputPath, column, outputTxt) {
  const XLSX = require("xlsx");
  const rows = readExcel(inputPath);
  if (!rows.length) return { error: "Excel 文件为空" };
  const headers = rows[0];
  const dataRows = rows.slice(1);
  const colIdx = resolveCol(headers, column);
  if (colIdx == null) return { need_column: true, columns: headers, preview: rows.slice(0, 6), message: column == null ? "无法自动判断 URL 列，请指定 --column 参数" : `找不到列 '${column}'` };

  const decoded = [];
  let success = 0;
  for (const row of dataRows) {
    const url = (row[colIdx] || "").trim();
    const d = url ? await decodeSingle(url) : FAIL;
    decoded.push(d);
    if (d !== FAIL) success++;
  }

  if (outputTxt) {
    fs.writeFileSync(outputTxt, decoded.join("\n"), "utf-8");
    return { total: dataRows.length, success, failed: dataRows.length - success, output_file: path.resolve(inputPath), output_txt: path.resolve(outputTxt) };
  }

  headers.push("解码结果");
  dataRows.forEach((r, i) => r.push(decoded[i]));
  const allRows = [headers, ...dataRows];
  const wb = XLSX.utils.book_new();
  const ws = XLSX.utils.aoa_to_sheet(allRows);
  XLSX.utils.book_append_sheet(wb, ws);
  XLSX.writeFile(wb, inputPath);
  return { total: dataRows.length, success, failed: dataRows.length - success, output_file: path.resolve(inputPath), output_txt: null };
}

// ── 入口 ──────────────────────────────────────────────

async function main() {
  const args = parseArgs(process.argv);
  if (!args.input) { console.log(JSON.stringify({ error: "用法: node batch_decode.js --input <文件> [选项]" })); process.exit(1); }

  const ext = path.extname(args.input).toLowerCase();
  let result;
  if (ext === ".txt") result = await processTxt(args.input, args.outputTxt);
  else if (ext === ".csv") result = await processCsv(args.input, args.column, args.outputTxt);
  else if (ext === ".xlsx" || ext === ".xls") result = await processExcel(args.input, args.column, args.outputTxt);
  else result = { error: `不支持的文件格式: ${ext}，支持 txt/csv/xlsx` };

  console.log(JSON.stringify(result));
  if (result.error) process.exit(1);
}

main();

File: skills/automation/search/qrcode-skills/scripts/batch_decode.py
"""
批量解码二维码 - 从 Excel/CSV/TXT 读取图片 URL，批量解码并写回结果。

用法:
  python scripts/batch_decode.py --input <文件> [选项]

选项:
  --column <列名或索引>   指定 URL 列（Excel/CSV 有效），不指定时自动检测
  --output-txt <文件>     单独输出结果到 TXT（按行分隔）

默认行为: 在原文件中新增一列写入解码结果。
TXT 输入时结果默认写到 <原文件名>_decoded.txt。

输出 JSON:
  成功: {"total": N, "success": N, "failed": N, "output_file": "...", "output_txt": "..."|null}
  需要用户指定列: {"need_column": true, "columns": [...], "preview": [...]}
  错误: {"error": "..."}
"""

import sys
import json
import os
import csv
import argparse
import tempfile
from pathlib import Path
from urllib.parse import quote

API_ENDPOINT = "https://api.2dcode.biz/v1/read-qr-code"
FAIL_PLACEHOLDER = "未解析到二维码"


def decode_single(image_source: str) -> str:
    """解码单个图片，返回内容文本或失败占位符。"""
    result = _try_zxing(image_source)
    if result:
        return result

    if _is_url(image_source):
        result = _decode_api_url(image_source)
    elif os.path.isfile(image_source):
        result = _decode_api_file(image_source)
    else:
        result = _decode_api_url(image_source)

    return result if result else FAIL_PLACEHOLDER


def _is_url(s: str) -> bool:
    return s.startswith("http://") or s.startswith("https://")


def _try_zxing(source: str) -> str | None:
    try:
        import zxingcpp
        from PIL import Image
    except ImportError:
        return None

    tmp_path = None
    try:
        if _is_url(source):
            import urllib.request
            suffix = Path(source.split("?")[0]).suffix or ".png"
            tmp = tempfile.NamedTemporaryFile(delete=False, suffix=suffix)
            urllib.request.urlretrieve(source, tmp.name)
            tmp.close()
            tmp_path = tmp.name
            img_path = tmp_path
        else:
            img_path = source

        img = Image.open(img_path)
        results = zxingcpp.read_barcodes(img)
        if results:
            return "; ".join(r.text for r in results)
        return None
    except Exception:
        return None
    finally:
        if tmp_path and os.path.exists(tmp_path):
            os.unlink(tmp_path)


def _decode_api_url(image_url: str) -> str | None:
    import urllib.request

    api_url = f"{API_ENDPOINT}?file_url={quote(image_url, safe='')}"
    try:
        with urllib.request.urlopen(api_url) as resp:
            data = json.loads(resp.read().decode())
        contents = data.get("data", {}).get("contents", [])
        if data.get("code") == 0 and contents:
            return "; ".join(contents)
        return None
    except Exception:
        return None


def _decode_api_file(file_path: str) -> str | None:
    import urllib.request
    import mimetypes
    import uuid

    boundary = uuid.uuid4().hex
    filename = os.path.basename(file_path)
    mime_type = mimetypes.guess_type(file_path)[0] or "application/octet-stream"

    with open(file_path, "rb") as f:
        file_data = f.read()

    body = (
        f"--{boundary}\r\n"
        f'Content-Disposition: form-data; name="file"; filename="{filename}"\r\n'
        f"Content-Type: {mime_type}\r\n\r\n"
    ).encode() + file_data + f"\r\n--{boundary}--\r\n".encode()

    req = urllib.request.Request(
        API_ENDPOINT,
        data=body,
        headers={"Content-Type": f"multipart/form-data; boundary={boundary}"},
    )
    try:
        with urllib.request.urlopen(req) as resp:
            data = json.loads(resp.read().decode())
        contents = data.get("data", {}).get("contents", [])
        if data.get("code") == 0 and contents:
            return "; ".join(contents)
        return None
    except Exception:
        return None


# ── 文件读写 ──────────────────────────────────────────────

def process_txt(input_path: str, output_txt: str | None):
    with open(input_path, "r", encoding="utf-8") as f:
        lines = [line.strip() for line in f if line.strip()]

    results = []
    success = 0
    for line in lines:
        decoded = decode_single(line)
        results.append(decoded)
        if decoded != FAIL_PLACEHOLDER:
            success += 1

    out_path = output_txt or str(Path(input_path).with_stem(Path(input_path).stem + "_decoded").with_suffix(".txt"))
    with open(out_path, "w", encoding="utf-8") as f:
        f.write("\n".join(results))

    return {
        "total": len(lines),
        "success": success,
        "failed": len(lines) - success,
        "output_file": os.path.abspath(out_path),
        "output_txt": os.path.abspath(out_path),
    }


def process_csv(input_path: str, column, output_txt: str | None):
    with open(input_path, "r", encoding="utf-8-sig", newline="") as f:
        rows = list(csv.reader(f))
    if not rows:
        return {"error": "CSV 文件为空"}

    headers = rows[0]
    data_rows = rows[1:]
    col_idx = _resolve_col(headers, column)

    if col_idx is None:
        return {
            "need_column": True,
            "columns": headers,
            "preview": rows[:6],
            "message": "无法自动判断 URL 列，请指定 --column 参数" if column is None else f"找不到列 '{column}'",
        }

    decoded_results = []
    success = 0
    for row in data_rows:
        url = row[col_idx].strip() if col_idx < len(row) else ""
        if url:
            decoded = decode_single(url)
        else:
            decoded = FAIL_PLACEHOLDER
        decoded_results.append(decoded)
        if decoded != FAIL_PLACEHOLDER:
            success += 1

    if output_txt:
        with open(output_txt, "w", encoding="utf-8") as f:
            f.write("\n".join(decoded_results))
        return {
            "total": len(data_rows),
            "success": success,
            "failed": len(data_rows) - success,
            "output_file": os.path.abspath(input_path),
            "output_txt": os.path.abspath(output_txt),
        }

    result_col = "解码结果"
    headers.append(result_col)
    for i, row in enumerate(data_rows):
        row.append(decoded_results[i])

    with open(input_path, "w", encoding="utf-8-sig", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(headers)
        writer.writerows(data_rows)

    return {
        "total": len(data_rows),
        "success": success,
        "failed": len(data_rows) - success,
        "output_file": os.path.abspath(input_path),
        "output_txt": None,
    }


def process_excel(input_path: str, column, output_txt: str | None):
    import openpyxl

    wb = openpyxl.load_workbook(input_path)
    ws = wb.active

    headers = [str(c.value) if c.value is not None else "" for c in ws[1]]
    col_idx = _resolve_col(headers, column)

    if col_idx is None:
        preview = []
        for row in ws.iter_rows(min_row=1, max_row=min(6, ws.max_row), values_only=True):
            preview.append([str(c) if c is not None else "" for c in row])
        wb.close()
        return {
            "need_column": True,
            "columns": headers,
            "preview": preview,
            "message": "无法自动判断 URL 列，请指定 --column 参数" if column is None else f"找不到列 '{column}'",
        }

    decoded_results = []
    success = 0
    for row_idx in range(2, ws.max_row + 1):
        cell_val = ws.cell(row=row_idx, column=col_idx + 1).value
        url = str(cell_val).strip() if cell_val else ""
        if url:
            decoded = decode_single(url)
        else:
            decoded = FAIL_PLACEHOLDER
        decoded_results.append(decoded)
        if decoded != FAIL_PLACEHOLDER:
            success += 1

    if output_txt:
        with open(output_txt, "w", encoding="utf-8") as f:
            f.write("\n".join(decoded_results))
        wb.close()
        return {
            "total": len(decoded_results),
            "success": success,
            "failed": len(decoded_results) - success,
            "output_file": os.path.abspath(input_path),
            "output_txt": os.path.abspath(output_txt),
        }

    result_col_idx = ws.max_column + 1
    ws.cell(row=1, column=result_col_idx, value="解码结果")
    for i, decoded in enumerate(decoded_results):
        ws.cell(row=i + 2, column=result_col_idx, value=decoded)

    wb.save(input_path)
    wb.close()

    return {
        "total": len(decoded_results),
        "success": success,
        "failed": len(decoded_results) - success,
        "output_file": os.path.abspath(input_path),
        "output_txt": None,
    }


# ── 列检测 ──────────────────────────────────────────────

def _resolve_col(headers: list[str], column) -> int | None:
    if column is not None:
        try:
            idx = int(column)
            return idx if 0 <= idx < len(headers) else None
        except ValueError:
            pass
        col_lower = str(column).lower().strip()
        for i, h in enumerate(headers):
            if h.strip().lower() == col_lower:
                return i
        return None

    keywords = ["url", "link", "image", "img", "图片", "链接", "网址", "二维码"]
    for i, h in enumerate(headers):
        h_lower = h.strip().lower()
        for kw in keywords:
            if kw in h_lower:
                return i
    if len(headers) == 1:
        return 0
    return None


# ── 入口 ──────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", required=True, help="输入文件路径")
    parser.add_argument("--column", default=None, help="指定 URL 列（列名或索引）")
    parser.add_argument("--output-txt", default=None, help="单独输出结果到 TXT")
    args = parser.parse_args()

    input_path = args.input
    ext = Path(input_path).suffix.lower()

    if ext == ".txt":
        result = process_txt(input_path, args.output_txt)
    elif ext == ".csv":
        result = process_csv(input_path, args.column, args.output_txt)
    elif ext in (".xlsx", ".xls"):
        result = process_excel(input_path, args.column, args.output_txt)
    else:
        result = {"error": f"不支持的文件格式: {ext}，支持 txt/csv/xlsx"}

    print(json.dumps(result, ensure_ascii=False))
    if "error" in result:
        sys.exit(1)


if __name__ == "__main__":
    main()
